statefulset-runner/controllers/pdb.go

```diff
@@ -0,0 +1,72 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const PdbMinAvailableInstances = "50%"
+
+type Updater struct {
+	client client.Client
+}
+
+func NewUpdater(client client.Client) *Updater {
+	return &Updater{
+		client: client,
+	}
+}
+
+func (c *Updater) Update(ctx context.Context, statefulSet *appsv1.StatefulSet) error {
+	if *statefulSet.Spec.Replicas > 1 {
+		return c.createPDB(ctx, statefulSet)
+	}
+
+	return c.deletePDB(ctx, statefulSet)
+}
+
+func (c *Updater) createPDB(ctx context.Context, statefulSet *appsv1.StatefulSet) error {
+	minAvailable := intstr.FromString(PdbMinAvailableInstances)
+
+	pdb := &policyv1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      statefulSet.Name,
+			Namespace: statefulSet.Namespace,
+			Labels: map[string]string{
+				LabelGUID:    statefulSet.Labels[LabelGUID],
+				LabelVersion: statefulSet.Labels[LabelVersion],
+			},
+		},
+		Spec: policyv1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+			Selector:     statefulSet.Spec.Selector,
+		},
+	}
+
+	if err := controllerutil.SetOwnerReference(statefulSet, pdb, scheme.Scheme); err != nil {
+		return errors.Wrap(err, "pdb-updated-failed-to-set-owner-ref")
+	}
+
+	err := c.client.Create(ctx, pdb)
+
+	if k8serrors.IsAlreadyExists(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "failed to create pod distruption budget")
```

Comment:
We should be consistent with other places in korifi where we wrap errors and use `fmt.Errorf("%w")`. We should not bring the unneeded dependency to `github.com/pkg/errors` from eirini.

Reply:
fixed